streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java

```diff
@@ -16,55 +16,139 @@
  */
 package org.apache.kafka.streams.processor.internals;
 
+import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.common.TopicPartition;
-import org.apache.kafka.streams.errors.StreamsException;
-import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.processor.StateStore;
 import org.apache.kafka.streams.processor.TaskId;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
 import java.util.Set;
 
 public interface Task {
 
-    void initializeMetadata();
-
-    /**
-     * Initialize the task and return {@code true} if the task is ready to run, i.e, it has no state stores
-     * @return true if this task has no state stores that may need restoring.
-     * @throws IllegalStateException If store gets registered after initialized is already finished
-     * @throws StreamsException if the store's change log does not contain the partition
-     */
-    boolean initializeStateStores();
+    long LATEST_OFFSET = -2L;
+
+    enum State {
+        CREATED {
+            @Override
+            public boolean hasBeenRunning() {
+                return false;
+            }
+        },
+        RESTORING {
+            @Override
+            public boolean hasBeenRunning() {
+                return false;
+            }
+        },
+        RUNNING {
+            @Override
+            public boolean hasBeenRunning() {
+                return true;
+            }
+        },
+        SUSPENDED {
+            @Override
+            public boolean hasBeenRunning() {
+                return true;
+            }
+        },
+        CLOSED {
+            @Override
+            public boolean hasBeenRunning() {
+                return true;
+            }
+        },
+        CLOSING {
+            @Override
+            public boolean hasBeenRunning() {
+                return true;
+            }
+        };
+
+        public abstract boolean hasBeenRunning();
```

Comment:
What is the semantic meaning of this method and why do we need it?

Reply:
The main purpose of this is to avoid calling `closeTopology` duplicated times if the state has already been transited to CLOSING.